drone/lint/lint.go

```diff
@@ -20,22 +27,57 @@ var Command = cli.Command{
 	},
 }
 
+type Flags struct {
+	Build  drone.Build
+	Netrc  drone.Netrc
+	Repo   drone.Repo
+	Stage  drone.Stage
+	System drone.System
+}
+
 func lint(c *cli.Context) error {
+	f := new(Flags)
+	f.Repo.Trusted = c.Bool("trusted")
+	var envs map[string]string
+
 	path := c.Args().First()
 	if path == "" {
 		path = ".drone.yml"
 	}
 
-	manifest, err := yaml.ParseFile(path)
+	rawsource, err := ioutil.ReadFile(path)
 	if err != nil {
 		return err
 	}
 
-	for _, resource := range manifest.Resources {
-		if err := linter.Lint(resource, c.Bool("trusted")); err != nil {
-			return err
+	// string substitution function ensures that string replacement variables are escaped and quoted if they contain newlines.
+	subf := func(k string) string {
+		v := envs[k]
+		if strings.Contains(v, "\n") {
+			v = fmt.Sprintf("%q", v)
 		}
+		return v
+	}
+	// evaluates string replacement expressions and returns an update configuration.
+	config, err := envsubst.Eval(string(rawsource), subf)
+	if err != nil {
+		return err
+	}
+	// parse and lint the configuration
+	manifest, err := manifest.ParseString(config)
+	if err != nil {
+		return err
+	}
+	// a configuration can contain multiple pipelines. get a specific pipeline resource for execution.
+	resource, err := resource.Lookup(f.Stage.Name, manifest)
```

Comment:
The runner command line tools only lint a single stage at a time.  This is by design since the runner only processes and lints a single stage at a time, and the command line tools are meant to help us isolate and test runner functionality.  They are intended to be used by runner developers, as opposed to end-users.

The drone command line tools, on the other hand, target end-users.  And for this reason we should probably lint all docker pipeline stages in the yaml.  I believe this can be achieved by doing something like this:

```Go
	manifest, err := manifest.ParseString(config)
	if err != nil {
		return err
	}

	for _, v := manifest.Resources {
		if v.Type == "docker" {
			stage, err := resource.Lookup(v.GetName(), manifest)
			if err != nil {
				return err
			}
			if err := linter.New().Lint(stage, &f.Repo); err != nil {
				return err
			}
		}
	}
```

The above code would only lint Docker pipelines for now, which is probably ok since this is how the current implementation works.

It would also be nice to check stage dependencies in the linter.

```

```Go
	manifest, err := manifest.ParseString(config)
	if err != nil {
		return err
	}

	// get list of stages
	stages := map[string]struct{}{}
	for _, v := manifest.Resources {
		stages[v.GetName()] = struct{}{}
	}

	// ensure all depends_on are valid
	for _, v := manifest.Resources {
		vv, ok := v.(resources. DependantResource)
		if !ok {
			continue
		}
		// get dependency list
		for _, dep := range vv.GetDependsOn() {
			if _, ok := stages[dep]; !ok {
				return fmt.Errorf("Pipeline stage %s declares invalid dependency %s")
			}
		}
	}
```

Reply:
i have made these changes.